pvlib/tests/test_tracking.py

```diff
@@ -486,3 +488,79 @@ def test_LocalizedSingleAxisTracker___repr__():
 
     assert localized_system.__repr__() == expected
+
+
+def test_calc_axis_tilt():
+    # expected values
+    expected_axis_tilt = 2.239  # [degrees]
+    expected_side_slope = 9.86649274360294  # [degrees]
+    expected = DATA_DIR / 'singleaxis_tracker_wslope.csv'
+    expected = pd.read_csv(expected, index_col='timestamp', parse_dates=True)
+    # solar positions
+    starttime = '2017-01-01T00:30:00-0300'
+    stoptime = '2017-12-31T23:59:59-0300'
+    lat, lon = -27.597300, -48.549610
+    times = pd.DatetimeIndex(pd.date_range(starttime, stoptime, freq='H'))
+    solpos = pvlib.solarposition.get_solarposition(times, lat, lon)
+    # singleaxis tracker w/slope data
+    system_plane = (77.34, 10.1149)
+    axis_azimuth = 0.0
+    max_angle = 75.0
+    # Note: GCR is relative to horizontal distance between rows
+    gcr = 0.33292759  # GCR = length / horizontal_pitch = 1.64 / 5 / cos(9.86)
+    # calculate tracker axis zenith
+    axis_tilt = tracking.calc_tracker_axis_tilt(
+        *system_plane, axis_azimuth=axis_azimuth)
+    assert np.isclose(axis_tilt, expected_axis_tilt)
+    # calculate side slope and relative rotation
+    side_slope = tracking.calc_system_tracker_side_slope(
+        axis_azimuth, axis_tilt, *system_plane)
+    assert np.isclose(side_slope, expected_side_slope)
+    sat = tracking.singleaxis(
+        solpos.apparent_zenith, solpos.azimuth, axis_tilt, axis_azimuth,
+        max_angle, backtrack=True, gcr=gcr, side_slope=side_slope)
+    np.testing.assert_allclose(
+        sat['tracker_theta'], expected['tracker_theta'], atol=1e-7)
+    np.testing.assert_allclose(sat['aoi'], expected['aoi'], atol=1e-7)
+    np.testing.assert_allclose(
+        sat['surface_azimuth'], expected['surface_azimuth'], atol=1e-7)
+    np.testing.assert_allclose(
+        sat['surface_tilt'], expected['surface_tilt'], atol=1e-7)
+
+
+def test_slope_aware_backtracking():
+    """
+    Test validation data set from https://www.nrel.gov/docs/fy20osti/76626.pdf
+    """
+    expected_data = np.array(
+        [('2019-01-01T08:00-0500',  2.404287, 122.79177, -84.440, -10.899),
+         ('2019-01-01T09:00-0500', 11.263058, 133.288729, -72.604, -25.747),
+         ('2019-01-01T10:00-0500', 18.733558, 145.285552, -59.861, -59.861),
+         ('2019-01-01T11:00-0500', 24.109076, 158.939435, -45.578, -45.578),
+         ('2019-01-01T12:00-0500', 26.810735, 173.931802, -28.764, -28.764),
+         ('2019-01-01T13:00-0500', 26.482495, 189.371536, -8.475, -8.475),
+         ('2019-01-01T14:00-0500', 23.170447, 204.13681, 15.120, 15.120),
+         ('2019-01-01T15:00-0500', 17.296785, 217.446538, 39.562, 39.562),
+         ('2019-01-01T16:00-0500',  9.461862, 229.102218, 61.587, 32.339),
+         ('2019-01-01T17:00-0500',  0.524817, 239.330401, 79.530, 5.490)],
+        dtype=[
+            ('Time', '<M8[h]'), ('ApparentElevation', '<f8'),
+            ('SolarAzimuth', '<f8'), ('TrueTracking', '<f8'),
+            ('Backtracking', '<f8')])
```

Comment:
Didn't know about numpy structured arrays.  Cool!

Reply:
FYI: AFAICT numpy structured arrays are the fastest form of tabular data there is, as fast as native numpy arrays, and faster than xarray, pandas, and numpy "record" arrays. Check out this [gist](https://nbviewer.jupyter.org/gist/mikofski/5fd38eda6f786c47a186c9d10b724eb9) that compares slicing tabular data by column and by row

# How fast to get a column
According to this very small sample only 8760, the structured array is fastest by a lot. All times in nanoseconds.

|index |np[:,n]  |np.T[n] |strc[key]|rec[key] |rec.attr|xray[key] |xray.attr|pd[key]  |pd.attr |
|----|----|----|----|----|----|-----|-----|----|----|
|36.7| 335| 301|175 |3820|4870|30700|84700|2370|6150|
|21.5| 352| 201|158 |2610|3400|26400|68100|1930|5040|

_Note_: for `np.T[n]` the array was transposed first, before timing the slice. New, numpy arrays are row-major (C order) by default. See [row-major](https://numpy.org/doc/stable/glossary.html#term-row-major)

# How fast to get a record?
According to this very small sample only 8760, the structured array is still the fastest!

|index[m] |np[m]  |np.T[:,m] |strc[m]|rec[m] |rec.attr|xray.isel(idx=m)  |xray[{idx:m}]|pd.iloc[m] |pd.loc[idx[m]] |
|----|----|----|----|----|----|------|-----|------|----|
|172 | 204| -- |154 |861 | n/a |213000| --  |300000| -- |
|138 | 195|242|145 |964 | n/a |222000|236000|133000| 767000 |

_Notes_:
* ditto for transposing before timnig
* Accessing a record from a record array using dot-notation in not-applicable
* Pandas and Xarray rows can be sliced by index or by integer. For pandas slicing by integer `pd.iloc[m]` was faster than by index, `pd.loc[idx]`.
* for slicing pandas, the index was known as a string before timing

Reply:
10x faster to get a column of data from that structured array than from a DataFrame. Should keep this in mind for improving performance of e.g. the solar position and clear sky calculations.